pkg/controller/database/cloudsql.go

```diff
@@ -0,0 +1,272 @@
+		return resource.ExternalCreation{}, errors.New(errNotCloudsql)
+	}
+	instance := cloudsql.GenerateDatabaseInstance(cr.Spec.ForProvider, meta.GetExternalName(cr))
+	_, err := c.db.Insert(c.projectID, instance).Context(ctx).Do()
+	return resource.ExternalCreation{}, errors.Wrap(resource.Ignore(gcp.IsErrorAlreadyExists, err), errInsertFailed)
+}
+
+func (c *cloudsqlExternal) Update(ctx context.Context, mg resource.Managed) (resource.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha2.CloudsqlInstance)
+	if !ok {
+		return resource.ExternalUpdate{}, errors.New(errNotCloudsql)
+	}
+	conn, err := c.updateRootCredentials(ctx, cr)
+	if err != nil {
+		return resource.ExternalUpdate{}, errors.Wrap(err, errUpdateRootFailed)
+	}
+	instance := cloudsql.GenerateDatabaseInstance(cr.Spec.ForProvider, meta.GetExternalName(cr))
+	_, err = c.db.Patch(c.projectID, meta.GetExternalName(cr), instance).Context(ctx).Do()
+	return resource.ExternalUpdate{ConnectionDetails: conn}, errors.Wrap(err, errPatchFailed)
+}
+
+func (c *cloudsqlExternal) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha2.CloudsqlInstance)
+	if !ok {
+		return errors.New(errNotCloudsql)
+	}
+	_, err := c.db.Delete(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+	return errors.Wrap(err, errDeleteFailed)
+}
+
+func (c *cloudsqlExternal) getConnectionDetails(ctx context.Context, cr *v1alpha2.CloudsqlInstance) (resource.ConnectionDetails, error) {
+	m := map[string][]byte{
+		v1alpha1.ResourceCredentialsSecretUserKey: []byte(cr.DatabaseUserName()),
+	}
+	s := &v1.Secret{}
+	name := types.NamespacedName{Name: cr.Spec.WriteConnectionSecretToReference.Name, Namespace: cr.Namespace}
+	err := c.kube.Get(ctx, name, s)
+	if resource.IgnoreNotFound(err) != nil {
+		return nil, err
+	}
+	if len(s.Data[v1alpha1.ResourceCredentialsSecretPasswordKey]) != 0 {
+		m[v1alpha1.ResourceCredentialsSecretPasswordKey] = s.Data[v1alpha1.ResourceCredentialsSecretPasswordKey]
+	}
+	// TODO(muvaf): There might be cases where more than 1 private and/or public IP address has been assigned. We should
+	// somehow show all addresses that are possible to use.
+	for _, ip := range cr.Status.AtProvider.IPAddresses {
+		if ip.Type == v1alpha2.PrivateIPType {
+			m[v1alpha2.PrivateIPKey] = []byte(ip.IPAddress)
+			// TODO(muvaf): we explicitly enforce use of private IP if it's available. But this should be configured
+			// by resource class or claim.
+			m[v1alpha1.ResourceCredentialsSecretEndpointKey] = []byte(ip.IPAddress)
+		}
+		if ip.Type == v1alpha2.PublicIPType {
+			m[v1alpha2.PublicIPKey] = []byte(ip.IPAddress)
+			if len(m[v1alpha1.ResourceCredentialsSecretEndpointKey]) == 0 {
+				m[v1alpha1.ResourceCredentialsSecretEndpointKey] = []byte(ip.IPAddress)
+			}
+		}
+	}
+	return m, nil
+}
+
+func (c *cloudsqlExternal) updateRootCredentials(ctx context.Context, cr *v1alpha2.CloudsqlInstance) (resource.ConnectionDetails, error) {
```

Comment:
It seems like this method:

1. Lists all users in the project.
1. Finds the one that represents the root database user.
1. If a password has not been set before (and saved to the connection secret), generates a new one.
1. Updates the root database user to use the new generated password.

How does all of this relate to the mysterious `RootPassword` field that is exposed by `sqladmin.DatabaseInstance` (but seemingly not documented by Google elsewhere)? If `RootPassword` is used to set the root user's password, could we avoid a lot of complexity by instead:

Inside `Create`:
1. Generating a root password
1. Creating the instance using the password generated in step 1
1. Returning `ConnectionDetails` containing the root username (as derived from the engine) and password generated in step 1

Then in `Observe`, returning `ConnectionDetails` containing the observed IP addresses and endpoint.

Keep in mind that the `ConnectionDetails` returned by each `ExternalClient` method are additive - `Create` can return half of the connection details, then `Observe` can return the other half and `PublishConnection` will take care of merging the two.

Reply:
> I can see this being a topic of discussion for other resources as well.

Agreed. I started thinking about this in https://github.com/crossplaneio/crossplane/pull/564, which we ended up abandoning. There's a whole bundle of complications that I don't think we're going to be able to work through as part of this push for v1beta1 resources, unfortunately.

In this specific instance I think we should keep it simple - set and publish the password at creation time and assume it won't be changed out of band. As you point out, if someone _really_ wanted to change their username or password they could always manually update the connection secret.

Reply:
Alright, that makes things easier. Though, somehow we should let users know that if they want to change the password they'd need to do it in the managed resource's secret not claim's secret since the former overrides the latter.